test/e2e_node: move getSampleDevicePluginPod to a test-only source file

hyangah opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

`image_list.go` is one of the files included in the non-test variant Go build list, but its `getSampleDevicePluginPod` function references `readDaemonSetV1OrDie` function defined in `device_plugin_test.go` which is included in the test variant Go build list only. (The file name is *_test.go).

As a result, "go build" fails with the undefined reference error.

In practice, that may not be an issue since k8s project contributors aren't meant to run go build on this package. However, tools that depend on `go build` to operate - e.g., gopls or govulncheck ./... - will report this as an error.

Fix this error and make test/e2e package pass `go build` by moving this file to also test-only source code (device_plugin_test.go)

#### Does this PR introduce a user-facing change?

NONE

ffromani commented:
/release-note none
/ok-to-test

ffromani commented:
/triage accepted
/priority backlog

hyangah commented:
/easycla

ffromani commented:
checking the dependencies lane we see that `image_list.go` is listed in a few places. Still the issue highlighted here is real. 
I'm now wondering if we can pivot a bit and move the `getSampleDevicePod` function in a new source file. Perhaps we can repurpose/fix `util_sampledevice.go` ?

hyangah committed:
e2e_node: move getSampleDevicePluginPod to device_plugin_test.go

image_list.go is one of the files included in the non-test variant Go build list, but its getSampleDevicePluginPod function references readDaemonSetV1OrDie function defined in device_plugin_test.go which is included in the test variant Go build list only. (The file name is *_test.go).

As a result, "go build" fails with the undefined reference error.

In practice, that may not be an issue since k8s project contributors aren't meant to run go build on this package. However, tools that depend on go build to operate - e.g., gopls or govulncheck ./... - will report this as an error.

Fix this error and make test/e2e package pass go build by moving this file to also test-only source code.

hyangah commented:
@ffromani I ended up moving `getSampleDevicePluginPod` to podresources_test.go instead.
I looked into a way to move this to one of those `util*` files including `util_sampledevice.go` and rename the file to `*_test.go`. However, the dependencies are so complex (for example, util_sampledevice.go defines multiple symbols referenced by non-test go files) and I concluded that require major refactoring and cleanup not suitable for a new contributor. 
I can move to the other file that calls this function (device_plugin_test.go) if you prefer.

MikeSpreitzer commented:
/lgtm

MikeSpreitzer commented:
/release-note-none

ffromani commented:
> @ffromani I ended up moving `getSampleDevicePluginPod` to podresources_test.go instead. I looked into a way to move this to one of those `util*` files including `util_sampledevice.go` and rename the file to `*_test.go`. However, the dependencies are so complex (for example, util_sampledevice.go defines multiple symbols referenced by non-test go files) and I concluded that require major refactoring and cleanup not suitable for a new contributor. I can move to the other file that calls this function (device_plugin_test.go) if you prefer.

I think this is fine for this PR. I agree we need to de-entangle the dependecy web and this task will require a much greater effort.

ffromani commented:
Hi @hyangah ! I'm happy with the direction here! I think we can simplify two things

1. we have 3 commits in the series, the first 2 are a rename and its revert. Can we drop both and narrow down the PR to just 1 commit?
2. `getSampleDevicePluginPod` is called in both `podresources_test.go` and `device_plugin_test.go`. It's an improvement to move it here from `image_list.go`, but I think it belong more to `device_plugin_test.go`. Besides the fact the sample device plugin fakes a device plugin (so it belongs more to device plugin area to the pod resources area), it's used in few testcases in device tests and in just one in podresources.

bart0sh commented:
@hyangah I agree with @ffromani here. Please move `getSampleDevicePluginPod` to the `device_plugin_test.go`
/hold
/assign @ffromani @bart0sh

hyangah commented:
Thanks. Moved getSampleDevicePluginPod to device_plugin_test.go, and squashed the commits.

ffromani commented:
/lgtm

thanks @hyangah for addressing the feedback so quickly!

bart0sh commented:
/retest

bart0sh commented:
/unhold

bart0sh commented:
/assign @klueska @mrunalp

ffromani commented:
Being this a e2e_node only change, perhaps @SergeyKanzhelev or @endocrimes could also help here?

SergeyKanzhelev reviewed (approved):
/lgtm
/approve

liggitt commented:
Thanks for this, I just fixed something similar in https://github.com/kubernetes/kubernetes/pull/119542

It looks like this PR fixes the only errors using govulncheck against release-1.27, release-1.26, and release-1.25. Would you mind opening cherry-picks of this PR to those branches?
